src/plot_api/plot_api.js

```diff
@@ -363,32 +365,49 @@ Plotly.plot = function(gd, data, layout, config) {
     });
 };
 
-
 function opaqueSetBackground(gd, bgColor) {
     gd._fullLayout._paperdiv.style('background', 'white');
-    Plotly.defaultConfig.setBackground(gd, bgColor);
+    defaultConfig.setBackground.dflt(gd, bgColor);
 }
 
 function setPlotContext(gd, config) {
-    if(!gd._context) gd._context = Lib.extendFlat({}, Plotly.defaultConfig);
-    var context = gd._context;
+    var keys, i;
 
-    if(config) {
-        Object.keys(config).forEach(function(key) {
-            if(key in context) {
-                if(key === 'setBackground' && config[key] === 'opaque') {
-                    context[key] = opaqueSetBackground;
-                }
-                else context[key] = config[key];
-            }
-        });
+    if(!gd._context) {
+        gd._context = {};
+        keys = Object.keys(defaultConfig);
 
-        // map plot3dPixelRatio to plotGlPixelRatio for backward compatibility
-        if(config.plot3dPixelRatio && !context.plotGlPixelRatio) {
-            context.plotGlPixelRatio = context.plot3dPixelRatio;
+        // init context
+        for(i = 0; i < keys.length; i++) {
+            var k = keys[i];
+            gd._context[k] = defaultConfig[k].dflt;
         }
     }
 
+    var context = gd._context;
+
+    config = Lib.isPlainObject(config) ? config : {};
+    keys = Object.keys(config);
+
+    function coerce(key, dflt) {
+        return Lib.coerce(config, context, defaultConfig, key, dflt);
+    }
+
+    // extend context with config
+    for(i = 0; i < keys.length; i++) {
+        coerce(keys[i]);
+    }
+
+    // 'opaque' is special value for setBackground
+    if(config.setBackground === 'opaque') {
+        context.setBackground = opaqueSetBackground;
+    }
+
+    // map plot3dPixelRatio to plotGlPixelRatio for backward compatibility
+    if(config.plot3dPixelRatio && !context.plotGlPixelRatio) {
+        context.plotGlPixelRatio = context.plot3dPixelRatio;
```

Comment:
shouldn't this happen to the input object before coercing, so it can be validated? Anyway if it stays here it would need to be `context.plotGlPixelRatio = config.plot3dPixelRatio;`